apis/python/src/tiledbsoma/tiledb_array.py

```diff
@@ -38,57 +52,31 @@ def schema(self) -> pa.Schema:
         """
         return get_arrow_schema_from_tiledb_uri(self.uri, self._ctx)
 
-    # tiledb.Array handle for [re]use while self is "open"
-    _open_tiledb_array: Optional[tiledb.Array] = None
-
-    def _sub_open(self) -> None:
-        assert self._open_mode in ("r", "w") and self._open_tiledb_array is None
-        self._open_tiledb_array = self._close_stack.enter_context(
-            tiledb.open(self._uri, mode=self._open_mode, ctx=self._ctx)
-        )
-
-    @property
-    def _tiledb_obj(self) -> tiledb.Array:
-        "get the open tiledb.Array handle"
-        assert self._open_tiledb_array is not None  # => not self.closed
-        return self._open_tiledb_array
-
-    def close(self) -> None:
-        self._open_tiledb_array = None
-        super().close()  # closes self._open_tiledb_array via self._close_stack
-
     def _tiledb_array_schema(self) -> tiledb.ArraySchema:
         """
         Returns the TileDB array schema. Not part of the SOMA API; for dev/debug/etc.
         """
-        with self._ensure_open():
-            return self._tiledb_obj.schema
+        return self._handle.reader.schema
 
-    def _tiledb_array_keys(self) -> Sequence[str]:
+    def _tiledb_array_keys(self) -> Tuple[str, ...]:
         """
         Return all dim and attr names.
         """
-        with self._ensure_open():
-            A = self._tiledb_obj
-            dim_names = [A.domain.dim(i).name for i in range(A.domain.ndim)]
-            attr_names = [A.schema.attr(i).name for i in range(A.schema.nattr)]
-            return dim_names + attr_names
+        return self._tiledb_dim_names() + self._tiledb_attr_names()
 
     def _tiledb_dim_names(self) -> Tuple[str, ...]:
         """
         Reads the dimension names from the schema: for example, ['obs_id', 'var_id'].
         """
-        with self._ensure_open():
-            A = self._tiledb_obj
-            return tuple([A.domain.dim(i).name for i in range(A.domain.ndim)])
+        arr = self._handle.reader
+        return tuple(arr.domain.dim(i).name for i in range(arr.domain.ndim))
 
-    def _tiledb_attr_names(self) -> List[str]:
+    def _tiledb_attr_names(self) -> Tuple[str, ...]:
         """
         Reads the attribute names from the schema: for example, the list of column names in a dataframe.
         """
-        with self._ensure_open():
-            A = self._tiledb_obj
-            return [A.schema.attr(i).name for i in range(A.schema.nattr)]
+        arr = self._handle.reader
+        return tuple(arr.schema.attr(i).name for i in range(arr.schema.nattr))
 
     def _soma_reader(
```

Comment:
I (naively?) expected the SOMAReader instance, which is really just a read-only TileDB handle, to be integrated into _handle.reader (or something similar).  At least for TileDB "array" objects, shouldn't all "handles" to the underlying array be managed in a consistent manner?

I'm thinking of stuff like the `flush(update_read=True)` call -- wouldn't we want the SOMAReader re-opened as well?

Reply:
I think until that is done, there are bugs aplenty to be found during the create process where handles get out of sync.  I put one trivial example into the comments, but I bet there are more due to the opportunity for the two separate read handles to get out of sync.

We really have a worst-case scenario right now - _three_ handles:
* TileDB-Py write handle
* TileDB-Py read handle for `.schema` and `.nonempty_domain`
* SOMAReader read handle for the rest of the read path
Ooof....

I wonder if @gspowley could make your life easier by providing all "read" path stuff in a single SOMAReader handle?

Reply:
I fixed a few spots where I was not properly checking for read mode when using the SOMAReader. This means that we have one of two situations:

1. Only a write handle, with an auxiliary read handle. The SOMAReader is never used.
2. A tiledb-py read handle, and a SOMAReader (temporary) read handle.

Because the SOMAReader is constructed with the same Context (technically, config) as the tiledb-py read handle, it will get the same view of the dataset as the read handle (pending ensuring that we sensibly set timestamps on everything).

Reply:
That is true for now, when we can only open the 'latest'.  As soon as @mlin lands his timestamp work, it will be quite easy to get them out of sync.  For the next PR :-)

Reply:
Right now a fresh `SOMAReader` is needed for each query, and some refactoring is needed in order to keep its TileDB handle open for multiple operations. @gspowley and I have discussed a couple possible approaches. Suggest this can be left to follow-up work.